Mathlib/Order/Notation.lean

```diff
@@ -54,11 +60,86 @@ class Inf (α : Type*) where
 
 attribute [ext] Min Max
 
-@[inherit_doc]
-infixl:68 " ⊔ " => Max.max
+/--
+The supremum/join operation: `x ⊔ y`. It is notation for `max x y`
+and should be used when the type is not a linear order.
+-/
+syntax:68 term:68 " ⊔ " term:69 : term
 
-@[inherit_doc]
-infixl:69 " ⊓ " => Min.min
+/--
+The infimum/meet operation: `x ⊓ y`. It is notation for `min x y`
+and should be used when the type is not a linear order.
+-/
+syntax:69 term:69 " ⊓ " term:70 : term
+
+macro_rules
+| `($a ⊔ $b) => `(Max.max $a $b)
+| `($a ⊓ $b) => `(Min.min $a $b)
+
+namespace Mathlib.Meta
+
+open Lean Meta PrettyPrinter Delaborator SubExpr Qq
+
+-- irreducible to not confuse Qq
+@[irreducible] private def linearOrderExpr (u : Level) : Q(Type u → Type u) :=
+  .const `LinearOrder [u]
+private def linearOrderToMax (u : Level) : Q((a : Type u) → $(linearOrderExpr u) a → Max a) :=
+  .const `LinearOrder.toMax [u]
+private def linearOrderToMin (u : Level) : Q((a : Type u) → $(linearOrderExpr u) a → Min a) :=
+  .const `LinearOrder.toMin [u]
+
+/--
+Return `true` if `LinearOrder` is imported and `inst` comes from a `LinearOrder e` instance.
+
+We use a `try catch` block to make sure there are no surprising errors during delaboration.
+-/
+private def hasLinearOrder (u : Level) (α : Q(Type u)) (cls : Q(Type u → Type u))
+    (toCls : Q((α : Type u) → $(linearOrderExpr u) α → $cls α)) (inst : Q($cls $α)) :
+    MetaM Bool := do
+  try
+    -- `isDefEq` may call type class search to instantiate `mvar`, so we need the local instances
+    withLocalInstances (← getLCtx).decls.toList.reduceOption do
+      let mvar ← mkFreshExprMVarQ q($(linearOrderExpr u) $α) (kind := .synthetic)
+      let inst' : Q($cls $α) := q($toCls $α $mvar)
+      isDefEq inst inst'
+  catch _ =>
+    -- For instance, if `LinearOrder` is not yet imported.
+    return false
+
+/-- Annotate the syntax `stx` with the go-to-def information of `target`. -/
+def withGoToDef (stx : Term) (target : Name) : DelabM Term := do
+  let stx ← annotateCurPos stx
+  let module := (← findModuleOf? target).getD (← getEnv).mainModule
+  let some range ← findDeclarationRanges? target | failure
+  let location := { module, range := range.selectionRange }
+  addDelabTermInfo (← getPos) stx (← getExpr) (location? := some location)
+  return stx
```

Comment:
@kmill, does this seem like an ok idea to you?

Reply:
Yes, I think that's fine. I might replace `failure` with `return stx` though, and move the `annotateCurPos` line until right before line 114.